poucave/main.py

```diff
@@ -47,41 +48,74 @@ def __init__(self):
             content = json.load(f)
         return web.json_response(content)
 
-    def checkpoint(self, project, name, description, module, ttl=None, params=None):
+    def checkpoint(
+        self,
+        project: str,
+        name: str,
+        description: str,
+        module: str,
+        ttl: Optional[int] = None,
+        params: Optional[Dict[str, Any]] = None,
+    ):
         ttl = ttl or config.DEFAULT_TTL  # ttl=0 is not supported.
-        params = params or {}
 
         mod = importlib.import_module(module)
         doc = (mod.__doc__ or "").strip()
         func = getattr(mod, "run")
 
+        conf_params = params or {}
+        url_params = getattr(mod, "URL_PARAMETERS", [])
+        types_url_params = {p: func.__annotations__[p] for p in url_params}
         exposed_params = getattr(mod, "EXPOSED_PARAMETERS", [])
-        filtered_params = {k: v for k, v in params.items() if k in exposed_params}
+        filtered_params = {k: v for k, v in conf_params.items() if k in exposed_params}
 
         infos = {
             "name": name,
             "project": project,
             "module": module,
             "description": description,
             "documentation": doc,
-            "parameters": filtered_params,
             "url": f"/checks/{project}/{name}",
+            "parameters": filtered_params,
         }
         self._checkpoints.append(infos)
 
         async def handler(request):
+            # Some parameters can be overriden in URL query.
+            try:
+                query_params = {
+                    # Convert submitted value to function param type.
+                    name: _type(request.query[name])
+                    for name, _type in types_url_params.items()
+                    if name in request.query
+                }
+                params = {**conf_params, **query_params}
+            except ValueError:
+                raise web.HTTPBadRequest()
+
+            # Some parameters are exposed in JSON response.
+            final_params = {k: v for k, v in params.items() if k in exposed_params}
+
             # Each check has its own TTL.
-            cache_key = f"{project}/{name}"
+            cache_key = f"{project}/{name}-" + ",".join(
+                f"{k}:{v}" for k, v in params.items()
+            )
```

Comment:
This uses the parameters from the conf as well. I think that's probably correct -- it means we ignore the cache if we change parameters -- but it also could mean that we have cached results from a previous version of the application if we restart with the same configuration.

Reply:
Yes, the thing is that the cache is in memory, so not persistent between restarts